### ISSUE TITLE
Drop 3.7 from travis and add 3.9 as "allow_failures"

### ISSUE THREAD

**Issue by aCampello** [wellcometrust/WellcomeML#295]
That would enable us to use some nice 3.8 tricks, for example assignment expressions. And would make us keep up with latest python.

**Comment by aCampello** [comment]
@nsorros ? 

And @ivyleavedtoadflax as a recent contributor :) any thoughts?

**Comment by ivyleavedtoadflax** [comment]
do it do it do it! Also if possible I would move over to github actions instead of travis.

**Comment by nsorros** [comment]
why remove 3.7 though?

**Comment by aCampello** [comment]
Because then we would be able to code in python 3.8, using the new features (and in the future in 3.9, and when `case/switch` lands in 3.10.... but I am getting ahead of myself!).

Note we can still use packages which are tested in <= 3.7 as dependencies, as those will be compatible.

**Comment by nsorros** [comment]
> Because then we would be able to code in python 3.8, using the new features (and in the future in 3.9, and when `case/switch` lands in 3.10.... but I am getting ahead of myself!).
> 
> Note we can still use packages which are tested in <= 3.7 as dependencies, as those will be compatible.

this will force people to use 3.8 which I think might be an unnecessary requirement. you can add an if statement to check for python version and implement it with the new and old way and we decide to switch to the new version we can keep one of the two.

**Comment by aCampello** [comment]
Hm, tbh I don't think 3.8 is a big ask. Note that `brew install python` already installs 3.9 by default, so we're 0.2 behind and very soon 0.3. 

But I hear you, in that most big dependencies that we use only forces `>=3.7` (for example [Sklearn 1.0](https://github.com/scikit-learn/scikit-learn#dependencies)). Perhaps we can be reactive and just continue coding in 3.7 for the sake of the package, and wait for them to force us to update.

**Comment by nsorros** [comment]
My rule of thumb is to use what spacy guys do. They current support python 3.6+ https://github.com/explosion/spaCy

But seriously though, we need to base that decision on uptake of certain python versions i.e. if a lot of users are still on 3.6 and 3.7 we should respect that. That said I understand some libraries need to take be "brave" as Apple says and drop support on earlier versions in order for users to upgrade.

![Screenshot 2021-06-02 at 3 10 08 PM](https://user-images.githubusercontent.com/4975761/120477605-a41fa900-c3b4-11eb-89c6-97f8be71f83a.png)
Source https://www.jetbrains.com/lp/python-developers-survey-2020/

Based on the above we would hurt ~30% of users.

I have not managed to find how other people are thinking which python version to support, other than the usage statistics above but another example of a tool supporting 3.6+ is `aws cli` https://aws.amazon.com/blogs/developer/announcing-the-end-of-support-for-python-3-4-and-3-5-in-the-aws-sdk-for-python-and-aws-cli-v1/

**Comment by aCampello** [comment]
Yeah, big libraries like aws cli have to have slower release cycles for various reasons. Google cloud SDK was `python 3.4` for a long time, even recently, but now they are `>3.5`. 

We need to require the maximum of our dependencies. So if we were to require `>= 3.6` we would have to pin sklearn and [numpy](https://pypi.org/project/numpy/) to older versions forever, which is a bit sad. I think even the new pandas requires `> 3.7`.

Also related tangentially, if we choose to test 3 python versions, our pipeline will have 6 concurrent runners. If we decide to go for windows, this can go up to 12 (unless we decide to only test 3.7 on windows, which is [Spacy's approach](https://github.com/explosion/spaCy/blob/ff5cf3606cf354c7639ce40b81cacc93bfb03583/azure-pipelines.yml#L38-L76)). Our travis is currently shared with the rest of Wellcome (RoRI and Explorify for example), so if we decide to support 3 versions in the test, I will have to set up a GitHub Actions on a private runner (probably AWS), or pay for a more expensive travis subscription, otherwise this can clash with other teams.
